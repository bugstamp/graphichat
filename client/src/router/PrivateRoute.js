import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Redirect } from 'react-router-dom';

import storage from '../storage';

class PrivateRoute extends Component {
  state = {
    render: false,
    auth: false,
  }

  componentDidMount() {
    const token = storage.token.get();

    if (token) {
      this.setState({
        render: true,
        auth: true,
      });
    } else {
      this.setState({
        render: true,
      });
    }
  }

  render() {
    const { render, auth } = this.state;
    const { children } = this.props;

    return (
      <Choose>
        <When condition={render}>
          <Choose>
            <When condition={!auth}>
              <Redirect to="login" />
            </When>
            <Otherwise>
              {children}
            </Otherwise>
          </Choose>
        </When>
        <Otherwise>{null}</Otherwise>
      </Choose>
    );
  }
}

PrivateRoute.propTypes = {
  children: PropTypes.oneOfType([PropTypes.element, PropTypes.node]).isRequired,
};

export default PrivateRoute;
