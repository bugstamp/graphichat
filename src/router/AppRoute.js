import React, { Fragment } from 'react';
// import PropTypes from 'prop-types';
import { Route, Redirect } from 'react-router-dom';

const AppRoute = ({
  component: Component,
  layout: Layout,
  privateRoute,
  ...rest
}) => (
  <Route
    {...rest}
    render={props => (
      <Choose>
        <When condition={privateRoute && !localStorage.getItem('chatkilla_tkn')}>
          <Redirect to="login" />
        </When>
        <When condition={!Layout}>
          <Fragment>
            <Component {...props} />
          </Fragment>
        </When>
        <Otherwise>
          <Layout>
            <Component {...props} />
          </Layout>
        </Otherwise>
      </Choose>
    )}
  />
);

export default AppRoute;
