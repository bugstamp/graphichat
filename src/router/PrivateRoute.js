import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Redirect } from 'react-router-dom';
import jwt from 'jsonwebtoken';

export const checkAuth = async (token) => {
  try {
    if (!token) {
      throw new Error('Token wasn\'t found');
    }
    const { data: { regStatus } } = await jwt.verify(token, process.env.TOKEN_SECRET);

    return regStatus;
  } catch (e) {
    throw e;
  }
};

class PrivateRoute extends Component {
  state = {
    render: false,
    auth: false,
    status: '',
  }

  async componentDidMount() {
    try {
      const token = localStorage.getItem('chatkilla_tkn');
      const status = await checkAuth(token);

      this.setState({
        render: true,
        auth: true,
        status,
      });
    } catch (e) {
      this.setState({
        render: true,
        auth: false,
        status: '',
      });
    }
  }

  render() {
    const { render, auth, status } = this.state;
    const { children } = this.props;

    return (
      <Choose>
        <When condition={render}>
          <Choose>
            <When condition={!auth}>
              <Redirect to="login" />
            </When>
            <When condition={auth && status === 'UNCOMPLETED'}>
              <Redirect to={`reg?step=2&token=${localStorage.getItem('chatkilla_tkn')}`} />
            </When>
            <Otherwise>
              {children}
            </Otherwise>
          </Choose>
        </When>
        <Otherwise>{null}</Otherwise>
      </Choose>
    );
  }
}

PrivateRoute.propTypes = {
  children: PropTypes.oneOfType([PropTypes.element, PropTypes.node]).isRequired,
};

export default PrivateRoute;
