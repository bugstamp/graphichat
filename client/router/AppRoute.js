import React from 'react';
import PropTypes from 'prop-types';
import { Route } from 'react-router-dom';

import PrivateRoute from './PrivateRoute';

const AppRoute = ({
  component: Component,
  layout: Layout,
  privateRoute,
  redirect,
  to,
  ...rest
}) => {
  const render = props => (
    <Layout>
      <Component {...props} />
    </Layout>
  );

  return (
    <Route
      {...rest}
      render={props => (
        <Choose>
          <When condition={redirect}>
            <Component to={to} />
          </When>
          <When condition={privateRoute}>
            <PrivateRoute>
              {render(props)}
            </PrivateRoute>
          </When>
          <Otherwise>
            {render(props)}
          </Otherwise>
        </Choose>
      )}
    />
  );
};

AppRoute.defaultProps = {
  privateRoute: false,
  layout: null,
  redirect: false,
  to: '',
};
AppRoute.propTypes = {
  component: PropTypes.oneOfType([PropTypes.element, PropTypes.node, PropTypes.func]).isRequired,
  layout: PropTypes.oneOfType([PropTypes.func, PropTypes.object]),
  privateRoute: PropTypes.bool,
  redirect: PropTypes.bool,
  to: PropTypes.string,
};

export default AppRoute;
