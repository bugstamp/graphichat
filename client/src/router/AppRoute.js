import React from 'react';
import PropTypes from 'prop-types';
import { Route } from 'react-router-dom';

import PrivateRoute from './PrivateRoute';

const AppRoute = ({
  component: Component,
  layout: Layout,
  privateRoute,
  redirect,
  to,
  ...rest
}) => {
  const renderLayout = props => (
    <Layout>
      <Component {...props} />
    </Layout>
  );

  const render = props => (
    <Choose>
      <When condition={redirect}>
        <Component to={to} />
      </When>
      <When condition={privateRoute}>
        <PrivateRoute>
          {renderLayout(props)}
        </PrivateRoute>
      </When>
      <Otherwise>
        {renderLayout(props)}
      </Otherwise>
    </Choose>
  );

  return (<Route {...rest} render={render} />);
};

AppRoute.defaultProps = {
  privateRoute: false,
  layout: null,
  redirect: false,
  to: '',
};
AppRoute.propTypes = {
  component: PropTypes.oneOfType([PropTypes.element, PropTypes.node, PropTypes.func]).isRequired,
  layout: PropTypes.oneOfType([PropTypes.func, PropTypes.object]),
  privateRoute: PropTypes.bool,
  redirect: PropTypes.bool,
  to: PropTypes.string,
};

export default AppRoute;
