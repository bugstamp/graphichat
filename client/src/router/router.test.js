import React from 'react';
import { MemoryRouter, Route } from 'react-router-dom';
import { shallow, mount } from 'enzyme';

import AppRoute from './AppRoute';
import PrivateRoute from './PrivateRoute';

const testComponent = () => (
  <div className="component" />
);

// eslint-disable-next-line
const testLayout = ({ children }) => (
  <div>{children}</div>
);

describe('test router', () => {
  describe('test AppRoute', () => {
    test('snapshot render', () => {
      const component = shallow(<AppRoute component={testComponent} />);

      expect(component).toMatchSnapshot();
    });

    test('Route render', () => {
      const routerWrapper = mount((
        <MemoryRouter initialEntries={['/']}>
          <AppRoute path="/" layout={testLayout} component={testComponent} />
        </MemoryRouter>
      ));

      expect(routerWrapper.find(Route).exists()).toEqual(true);
      expect(routerWrapper.find(Route).prop('render')).toBeDefined();
    });

    test('pass props to Route', () => {
      const routerWrapper = mount((
        <MemoryRouter initialEntries={['/']}>
          <AppRoute path="/" layout={testLayout} component={testComponent} testProps="testProps" />
        </MemoryRouter>
      ));

      expect(routerWrapper.find(Route).prop('testProps')).toEqual('testProps');
    });

    test('redirect render', () => {
      const routerWrapper = mount((
        <MemoryRouter initialEntries={['/']}>
          <AppRoute path="/" layout={testLayout} component={testComponent} redirect />
        </MemoryRouter>
      ));

      expect(routerWrapper.find(testComponent).exists()).toEqual(true);
      expect(routerWrapper.find(testLayout).exists()).toEqual(false);
    });

    test('layout render', () => {
      const routerWrapper = mount((
        <MemoryRouter initialEntries={['/']}>
          <AppRoute path="/" layout={testLayout} component={testComponent} />
        </MemoryRouter>
      ));

      expect(routerWrapper.find(testComponent).exists()).toEqual(true);
      expect(routerWrapper.find(testLayout).exists()).toEqual(true);
    });

    test('privateRoute render', () => {
      const routerWrapper = mount((
        <MemoryRouter initialEntries={['/']}>
          <AppRoute path="/" layout={testLayout} component={testComponent} privateRoute />
        </MemoryRouter>
      ));

      expect(routerWrapper.find(PrivateRoute).exists()).toEqual(true);
    });
  });
});
