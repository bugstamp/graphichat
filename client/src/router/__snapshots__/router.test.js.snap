// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test router test AppRoute snapshot render 1`] = `
<Route
  render={[Function]}
/>
`;
